File: LeetCode/April Challenge/18th Day/Day_18.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        vector<vector<int>> temp(grid);
        if(temp.size()==0)
            return 0;
        for(int i=0;i<temp.size();i++)
        {
            for(int j=0;j<temp[i].size();j++)
            {
                if(i==0 && j>0)
                    temp[i][j]+=temp[i][j-1];
                else
                {
                    if(j==0 && i>0)
                        temp[i][j]+=temp[i-1][j];
                    else
                        if(i>0 && j>0)
                             temp[i][j]+=min(temp[i-1][j],temp[i][j-1]);
                }
            }
        }
        return temp[temp.size() -1][temp[0].size()-1];
    }
};

File: LeetCode/April Challenge/18th Day/Day_18.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int: 
        n, m = len(grid), len(grid[0])
        
        dp = grid # Initialize Memoization table
        
        # Base Case For Row
        for j in range(1, m):
            dp[0][j] = grid[0][j] + dp[0][j - 1]
            
        # Base Case For Col
        for i in range(1, n):
            dp[i][0] = grid[i][0] + dp[i-1][0]
            
        # General Case.
        for i in range(1, n):
            for j in range(1, m):
                dp[i][j] = grid[i][j] + min(dp[i][j-1], dp[i - 1][j])
                
        return dp[n - 1][m - 1]
    
